[libc++] Implement `views::join_with`

JMazurkiewicz opened the pull request:
* Implement "P2441R2 `views::join_with`" (https://wg21.link/P2441R2), closes #105185
* Implement LWG4074 (https://wg21.link/LWG4074), closes #105346
* Complete implementation of "P2711R1 Making multi-param constructors of views explicit" (https://wg21.link/P2711R1), closes #105252
* Complete implementation of "P2770R0 Stashing stashing iterators for proper flattening" (https://wg21.link/P2770R0), closes #105250

Zingam commented on libcxx/docs/Status/Cxx23Papers.csv:
```diff
@@ -104,9 +104,9 @@
 "","","","","","",""
 "`P0290R4 <https://wg21.link/P0290R4>`__","LWG", "``apply()`` for ``synchronized_value<T>``","February 2023","","","|concurrency TS|"
-"`P2770R0 <https://wg21.link/P2770R0>`__","LWG", "Stashing stashing ``iterators`` for proper flattening","February 2023","","","|ranges|"
+"`P2770R0 <https://wg21.link/P2770R0>`__","LWG", "Stashing stashing ``iterators`` for proper flattening","February 2023","|Partial| [#note-P2770R0]_","","|ranges|"
```
I was looking into implementing `join_with` but since you arelady did such a good job, are you planning also to finish the `join_view` part as that view is marked experimental as it hasn't complete these changes?

JMazurkiewicz replied:
Yes, I'm working on it too.

Zingam replied:
@JMazurkiewicz Great! You should probably update the status page that you are working on these items:
https://github.com/llvm/llvm-project/blob/main/libcxx/docs/Status/RangesViews.csv
https://libcxx.llvm.org/Status/Ranges.html

Although for some reason it doesn't get rendered right now. I guess it is a CI issue. BTW. You can use a VSCode extension to check if rst files are rendered correctly.

ldionne commented:
@JMazurkiewicz Please ping us when this is ready to review and make it a non-draft.

JMazurkiewicz committed:
[libc++] Implement `views::join_with`

* Implement "P2441R2 `views::join_with`" (https://wg21.link/P2441R2)
* Complete implementation of "P2711R1 Making multi-param constructors of views explicit" (https://wg21.link/P2711R1)
* Complete implementation of "P2770R0 Stashing stashing iterators for proper flattening" (https://wg21.link/P2770R0)

H-G-Hristov commented:
I guess this cool feature can be finalized now :)

Zingam commented:
@JMazurkiewicz We have guidelines for applying `[[nodiscard]]` since recently (in case you missed the discussion). I think you can/ or should/ apply `[[nodiscard]]` wherever it is relevant now (operators, etc.).

huixie90 commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+
+  static constexpr bool _UseOuterItCache = !forward_range<_View>;
+  using _OuterItCache = _If<_UseOuterItCache, __non_propagating_cache<iterator_t<_View>>, __empty_cache>;
+  _LIBCPP_NO_UNIQUE_ADDRESS _OuterItCache __outer_it_;
```
do we have a test for this `no_unique_address`? that is, if we don't need the cache, the test should check that we don't waste the memory

JMazurkiewicz replied:
Yes we do, here: [libcxx/test/libcxx/ranges/range.adaptors/range.join.with/range.join.with.iterator/no_unique_address.compile.pass.cpp](https://github.com/JMazurkiewicz/llvm-project/blob/libcxx/ranges/join_with/libcxx/test/libcxx/ranges/range.adaptors/range.join.with/range.join.with.iterator/no_unique_address.compile.pass.cpp)

huixie90 commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+
+  static constexpr bool _UseInnerCache = !is_reference_v<_InnerRng>;
+  using _InnerCache = _If<_UseInnerCache, __non_propagating_cache<remove_cvref_t<_InnerRng>>, __empty_cache>;
+  _LIBCPP_NO_UNIQUE_ADDRESS _InnerCache __inner_;
```
same as above. do we have a test for the `no_unique_address`?

JMazurkiewicz replied:
Yes, in the same file as above.

huixie90 commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr _OuterIter& __get_outer() {
```
A note to myself, reviewed this header until this line

huixie90 reviewed (changes requested):
Thanks for the patch. it looks really good. I started looking at the join_with_view.h and has not finished the header yet.

JMazurkiewicz commented:
> Gentle ping on this PR -- do you think you'll have time to pick it up or should we find someone else to get it through the finish line?

I'll take care of this PR again, starting next week.

frederick-vs-ja commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+  using _OuterIterType                                 = _If<_OuterIterPresent, _OuterIter, std::__empty>;
+  _LIBCPP_NO_UNIQUE_ADDRESS _OuterIterType __outer_it_ = _OuterIterType();
+
+  variant<_PatternIter, _InnerIter> __inner_it_;
```
Per current standard wording, some operations are required to throw `bad_variant_access`. It's unclear whether this is intended, and [LWG4059](https://cplusplus.github.io/LWG/issue4059) intends to make such operations UB when `valueless_by_exception`.

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Add `[[nodiscard]]` to private functions

JMazurkiewicz committed:
Add third template argument in `__join_with_view_iterator_category`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1779496537

JMazurkiewicz committed:
Add missing <includes>

JMazurkiewicz committed:
Clang 17 is no longer supported

JMazurkiewicz committed:
Implement LWG-4074

JMazurkiewicz committed:
Usa `[[no_unique_address]]` in `join_with_view::sentinel`

JMazurkiewicz committed:
Cleanup `[[no_unique_address]]` tests

Related comments:
* https://github.com/llvm/llvm-project/pull/65536#discussion_r1779740468
* https://github.com/llvm/llvm-project/pull/65536#discussion_r1779740529
* https://github.com/llvm/llvm-project/pull/65536#discussion_r1780034366

JMazurkiewicz committed:
Test when result of `operator*` is type different from `range_reference_t<InnerRng>` and `range_reference_t<Pattern>`

Related comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1780064183

JMazurkiewicz committed:
Various cleanups

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Fix formatting

JMazurkiewicz committed:
Fix `example.pass.cpp` test

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+
+  _LIBCPP_NODISCARD _LIBCPP_HIDE_FROM_ABI constexpr auto begin() const
+    requires forward_range<const _View> && forward_range<const _Pattern> &&
+             is_reference_v<range_reference_t<const _View>> && input_range<range_reference_t<const _View>>
```
Addressed here: https://github.com/llvm/llvm-project/pull/65536/commits/67784267be6c04786ff9bed56026e2b63134b2f0

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+
+  static constexpr bool _OuterIterPresent              = forward_range<_Base>;
+  using _OuterIterType                                 = _If<_OuterIterPresent, _OuterIter, std::__empty>;
+  _LIBCPP_NO_UNIQUE_ADDRESS _OuterIterType __outer_it_ = _OuterIterType();
```
Yes, here: https://github.com/JMazurkiewicz/llvm-project/blob/libcxx/ranges/join_with/libcxx/test/libcxx/ranges/range.adaptors/range.join.with/range.join.with.iterator/no_unique_address.compile.pass.cpp

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,456 @@
+  }
+
+  _LIBCPP_NODISCARD _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator*() const {
+    using __reference = common_reference_t<iter_reference_t<_InnerIter>, iter_reference_t<_PatternIter>>;
```
I've added such test in this commit: https://github.com/llvm/llvm-project/pull/65536/commits/e55c077058a39d1f123a1a8282c77b972ab9efad

JMazurkiewicz committed:
Fix `19.rst` file

JMazurkiewicz committed:
Enable `iterator/no_unique_address.compile.pass.cpp` test for MSVC

JMazurkiewicz committed:
Fix formatting (again)

JMazurkiewicz committed:
Remove empty `test_end` function

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897977473

JMazurkiewicz committed:
Test explicitness of default constructor of `join_with_view`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897974523

JMazurkiewicz committed:
Test explicitness of `join_with_view::iterator(iterator<!Const>)` constructor

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897426169

JMazurkiewicz committed:
Test explicitness of default constructor of `join_with_view::iterator`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897425622

JMazurkiewicz committed:
Parametrise `test_pre_increment` on `ref-is-glvalue` being `true` or `false`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897428139

JMazurkiewicz committed:
Add extra test for `iter_move` (make sure it calls underlying `iter_move`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897430717

JMazurkiewicz committed:
Remove GCC workaround from `increment.pass.cpp`

JMazurkiewicz committed:
Add extra test for `iter_swap` (make sure it calls underlying `iter_swap`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897431006

JMazurkiewicz committed:
Make sure `CIter` and `Iter` are different types (`ctor.not_const.pass.cpp`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897401275

JMazurkiewicz committed:
Make sure `CIter` and `Iter` are different types (`eq.pass.cpp`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897429572

JMazurkiewicz committed:
Make sure `CSent` and `Sent` are different types (`ctor.not_const.pass.cpp`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897431354

JMazurkiewicz committed:
Make sure `CSent` and `Sent` are different types (`eq.pass.cpp`)

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897431571
Drive-by: make use of new aliases

JMazurkiewicz committed:
Add missing `<array>` include in `iter_swap.pass.cpp`

JMazurkiewicz committed:
Fix `sentinel::operator==`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r1897405679

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Fix formatting

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.view/ctor.default.pass.cpp:
```diff
@@ -0,0 +1,73 @@
+
+// <ranges>
+
+// join_with_view()
```
Added: https://github.com/llvm/llvm-project/pull/65536/commits/4ab1ee25812e61e967870d6c7a9c5b14877cfa8d

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.sentinel/eq.pass.cpp:
```diff
@@ -0,0 +1,101 @@
+
+// template<bool OtherConst>
+//   requires sentinel_for<sentinel_t<Base>, iterator_t<maybe-const<OtherConst, V>>>
+// friend constexpr bool operator==(const iterator<OtherConst>& x, const sentinel& y);
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/96eae0b02b289e7eca0c7d5ecd764b7c01a26b16

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.sentinel/ctor.non_const.pass.cpp:
```diff
@@ -0,0 +1,71 @@
+    std::ranges::join_with_view jwv(vec, 0);
+    using JWV = decltype(jwv);
+    static_assert(!std::ranges::common_range<JWV>);
+    using CSent = std::ranges::sentinel_t<const JWV>;
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/d52bdbf9d5f6f67afb0417a47993e149a3f647e4

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/eq.pass.cpp:
```diff
@@ -0,0 +1,256 @@
+    using JWV     = std::ranges::join_with_view<std::ranges::owning_view<V>, std::ranges::owning_view<Pattern>>;
+
+    using Iter  = std::ranges::iterator_t<JWV>;
+    using CIter = std::ranges::iterator_t<const JWV>;
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/52ad025f54b94ef7e2c8bcdb213a1f6ce0a02363

> I don't see it in the spec. I suppose it works because of the implicit conversion constructor

You're right, it works thanks to that constructor.

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/ctor.not_const.pass.cpp:
```diff
@@ -0,0 +1,104 @@
+
+// <ranges>
+
+// constexpr iterator(iterator<!Const> i)
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/bf1eeacad13967a404d24f777ef104d4105f41f8

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/ctor.default.pass.cpp:
```diff
@@ -0,0 +1,77 @@
+
+// <ranges>
+
+// iterator() = default;
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/3771be94b248472d02dc1459d3252d5e821937b6

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/iter_move.pass.cpp:
```diff
@@ -0,0 +1,347 @@
+
+// <ranges>
+
+// friend constexpr decltype(auto) iter_move(const iterator& x);
```
Yes we can! Here: https://github.com/llvm/llvm-project/pull/65536/commits/42ce9c27a7b1b2ddcfb66e815bc7df3c30ed5af3

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/iter_swap.pass.cpp:
```diff
@@ -0,0 +1,78 @@
+
+// <ranges>
+
+// friend constexpr void iter_swap(const iterator& x, const iterator& y);
```
Done! https://github.com/llvm/llvm-project/pull/65536/commits/19e4fa2857b83a67475d9faa94e93636055329f1

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/ctor.not_const.pass.cpp:
```diff
@@ -0,0 +1,104 @@
+    auto it = jwv.begin();
+    assert(*it == 1);
+
+    using CIter      = std::ranges::iterator_t<const decltype(jwv)>;
```
Added here: https://github.com/llvm/llvm-project/pull/65536/commits/14471bc8aba671477e20991878d039389d5fa0f0

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/increment.pass.cpp:
```diff
@@ -0,0 +1,320 @@
+
+#endif // !defined(TEST_COMPILER_GCC)
+
+  { // Only last element of `V` is not empty. `Pattern` is empty. `V` models input range and `ref-is-glvalue` is false.
```
Here is my attempt: https://github.com/llvm/llvm-project/pull/65536/commits/04942c07124c530392cf0c7b7ad712c7efa93190

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.view/end.pass.cpp:
```diff
@@ -0,0 +1,234 @@
+  }
+}
+
+constexpr void test_end() {}
```
It's not, I forgot to remove it earlier (https://github.com/llvm/llvm-project/pull/65536/commits/fb5ab018b425d2a686f32bc77dfe90b0a8ee2bcd).

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,453 @@
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator(__iterator<!_Const> __i)
+    requires _Const && convertible_to<iterator_t<_View>, _OuterIter> &&
+                 convertible_to<iterator_t<_InnerRng>, _InnerIter> && convertible_to<iterator_t<_Pattern>, _PatternIter>
+      : __parent_(__i.__parent_), __outer_it_(std::move(__i.__outer_it_)) {
```
> I think there is something wrong. `__iterator<false>` is not a friend of `__iterator<true>`, so in theory this should not compile because of accessing private members.
> 
> This probably indicates that this function is not tested

I think that everything is OK here. Even though `__iterator<false>` is not a friend of `__iterator<true>`, its members can still be accessed thanks to `friend join_with_view` declaration. Per [[class.friend]/2](http://eel.is/c++draft/class.friend):

> Declaring a class to be a friend implies that private and protected members of the class granting friendship can be named in the base-specifiers and member declarations of the befriended class.

`__iterator<false>` declares `join_with_view` to be a friend, which means that member declarations of it (`__iterator<true>` and `__sentinel<?>`) can access iterator's private members.

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,453 @@
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __sentinel(__sentinel<!_Const> __s)
+    requires _Const && convertible_to<sentinel_t<_View>, sentinel_t<_Base>>
+      : __end_(std::move(__s.__end_)) {}
```
I think that [[class.friend]/2](http://eel.is/c++draft/class.friend#2) applies here too.

JMazurkiewicz commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,453 @@
+    requires sentinel_for<sentinel_t<_Base>, iterator_t<__maybe_const<_OtherConst, _View>>>
+  [[__nodiscard__]] _LIBCPP_HIDE_FROM_ABI friend constexpr bool
+  operator==(const __iterator<_OtherConst>& __x, const __sentinel& __y) {
+    return __x.__get_outer() == __y.__end_;
```
You are right here, `operator==` is a friend of `sentinel` only, so it should not be able to access `iterator`'s private members. Unluckily, both clang and gcc accept this incorrect code.

Fixed here: https://github.com/llvm/llvm-project/pull/65536/commits/9760ec3ca84fde34aa75ea469d618af84532ed7f

JMazurkiewicz committed:
Clarify some comments in `test_post_increment`

JMazurkiewicz commented on libcxx/test/std/ranges/range.adaptors/range.join.with/range.join.with.iterator/increment.pass.cpp:
```diff
@@ -0,0 +1,320 @@
+// constexpr iterator& operator++();
+// constexpr void operator++(int);
+// constexpr iterator operator++(int)
+//   requires ref-is-glvalue && forward_iterator<OuterIter> &&
```
Yes, we test them by verifying the return type of `operator++(int)` depending on underlying range in `test_post_increment`. I've added extra comments to make it clear here: https://github.com/llvm/llvm-project/pull/65536/commits/d3c014f8261bbc3a9dc6b5203254452f9ea851b3

Zingam commented:
@JMazurkiewicz At the top you mention that this PR completed several papers. Do you need to add additional GitHub issues to the list that this PR closes/fixes/?

JMazurkiewicz committed:
Use `_LIBCPP_NODEBUG` attribute

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
[NFC] Use `// REQUIRES: std-at-least-c++23`

Instead of `// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20`

JMazurkiewicz committed:
More `_LIBCPP_NODEBUG`

JMazurkiewicz committed:
Fix unstable ABI tests

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Target LLVM 21

JMazurkiewicz committed:
Trigger CI due to GitHub "incident"

https://www.githubstatus.com/incidents/nm83zrdky73y

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Add links to GitHub in `21.rst`

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz committed:
Don't use `std::is_trivial` (deprecated in C++26)

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

JMazurkiewicz commented:
Ping @huixie90

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

huixie90 reviewed (approved):
LGTM with some nits. Thanks very much for the patch!

Mick235711 commented on libcxx/docs/ReleaseNotes/21.rst:
```diff
@@ -47,6 +47,9 @@ Implemented Papers
 - P2897R7: ``aligned_accessor``: An mdspan accessor expressing pointer over-alignment (`Github <https://github.com/llvm/llvm-project/issues/118372>`__)
 - P3247R2: Deprecate the notion of trivial types (`Github <https://github.com/llvm/llvm-project/issues/118387>`__)
+- P2441R2: ``views::join_with`` (`Github <https://github.com/llvm/llvm-project/issues/105185>`__)
+- P2711R1: Making multi-param constructors of ``views`` ``explicit`` (`Github <https://github.com/llvm/llvm-project/issues/105252>`__)
```
I think the rest of P2711 is [already done](https://reviews.llvm.org/D144822?id=500542) in libc++ 17.

Zingam replied:
I confirm. I think it was my first contribution.

philnik777 commented on libcxx/include/__ranges/join_with_view.h:
```diff
@@ -0,0 +1,461 @@
+
+  [[__nodiscard__]] _LIBCPP_HIDE_FROM_ABI constexpr _View base() && { return std::move(__base_); }
+
+  [[__nodiscard__]] _LIBCPP_HIDE_FROM_ABI constexpr auto begin() {
```
This should use the non-ugly spelling. Re. policy, we have https://libcxx.llvm.org/CodingGuidelines.html#apply-nodiscard-where-relevant. In this case I think it's fine. This would most likely fall under "only observe a value" or "is most likely a misuse of the function". While it's technically possible that someone might call `begin()` to cache something, this would definitely require a comment in code, and having to add a void cast seems fine to me in that case.

ldionne commented:
Sorry for the churn about the `pending-ci` tag -- I missed that there were still a few comments to resolve before merging.

JMazurkiewicz committed:
`is_reference_v<range_reference_t<_Base>>` => `is_reference_v<_InnerBase>`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r2081562485

JMazurkiewicz committed:
`[[__nodiscard__]]` => `[[nodiscard]]`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r2081531902

JMazurkiewicz committed:
Add static member function in `__sentinel` to access priavate member of `__iterator`

Comment: https://github.com/llvm/llvm-project/pull/65536#discussion_r2081747606

JMazurkiewicz committed:
Fix `iter_swap.pass.cpp` (friendship issues)

JMazurkiewicz committed:
Add `static_cast<bool>` in `end.pass.cpp` to avoid truncation warnings with MSVC

JMazurkiewicz committed:
Update synopsis in `<ranges>` (LWG-4074)

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/ranges/join_with

frederick-vs-ja reviewed (approved):
LGTM. The CI failures are probably unrelated.

huixie90 commented:
Merging this as comments are addressed. CI is green except for msan, which is an existing issue on main and unrelated to this patch
